department-of-veterans-affairs/caseflow
# Caseflow 1857 prepopulate decision document id and fix modifications

Resolves [Caseflow 1857](https://vajira.max.gov/browse/CASEFLOW-1857)

### Description
From the Jira ticket:  As a user who has access to edit the decision document ID, I want the edit modal to pre-populate the text box with the existing decision document ID, so that I can edit the ID without having to retype it. 

### Acceptance Criteria
- [ ] the user no longer sees placeholder text on the edit modal text box
- [ ] the user see the existing decision document ID in the modal text box
- [ ] the user is able to edit the existing decision document ID in the modal text box
- [ ] the user is unable to save the edit if the decision document ID entered does not meet the validation requirements "Document ID of type Draft Decision must be in one of these formats: 12345-12345678 or 12345678.123 or 12345678.1234"
- [ ] disable submit if no change in text box has been made.

### Changes Made:
- [ ] Existing bug where there were several work product codes for legacy appeals that didn't exist in the validator. So they were failing to match up with the right error message for formatting the number.
- [ ] In the modal, removed the placeholder text and replaced that with the existing document id that is editable.
- [ ] Because the text field in the modal was linked to the decision document id field in the CaseTitleDetails section, when the user updated the ID in the modal, it was also being updated at the same time in the decision document ID field on the main page. This implied to the user that the field was updated when in fact it was not yet saved to the backend. Upon saving, the success message displayed, but it was not completely saving.  Added a call to the Redux store during the save process so that it would update appealDetails too.
- [ ] Added a check for the disable button to not enable if the user entered the existing document id


### Testing Plan
1. Rspec for spec/feature/queue/case_details/updated_document_id_spec.rb.  No changes were made to this file. Existing validations check AMA and Legacy Appeals for proper formatting, displaying and error if the formatting is wrong, allowing editing, etc. 
2. Manual testing:
- Login in as BVAAABSHIRE and go the Queue page.
- Choose a legacy appeal that has an existing Decision Document ID
- Make note of the existing document ID and click on the Edit button.
- The text field inside the modal should have contain the existing document ID and it should be editable.
- The "SAVE" button should be disabled until the value of the field has changed. If you change it and revert it back to the same number, the SAVE button should enable and disable.
- Edit the value in the text field and click on Save.  If the value was the right format, it will succeed. If it doesn't, an error will display.
- A successful save shows the updated document id in the Decision Document ID field in the CaseTitleDetails section.
- To test an AMA appeal, login as Attorney Casper, select an AMA case from his Queue, use the dropdown for "Decision Ready for Review", and go through the steps to setup the original Document ID.
- Once complete, the success message will says it's been assigned to judge XXX (this was BVAAABshire for me).
- Switch to BVAAABshire's Queue, find the case, and edit the Document ID.  Notice the same activity as the legacy appeal.

### User Facing Changes - Screenshot of the Error message
<img width="487" alt="error_for_wrong_format" src="https://user-images.githubusercontent.com/23639439/131927430-06ae2d42-ed98-40e7-bf3e-f58104fbde0b.png">

### Storybook Story
*For Frontend (Presentational) Components*
- [ ] Not related to this user story, but made a Storybook fix for SubstituteAppellantReview. It was crashing post a PR that updated 2 name fields that were not updated in Storybook. Could address separately, but it doesn't affect prod code.
